File: apps/api/src/services/billing/webhooks.py
"""
Stripe webhook event handlers.

All handlers are idempotent — they check for existing records before creating
new ones to safely handle Stripe's at-least-once delivery guarantee.
"""

import logging
from datetime import datetime

import stripe
from sqlmodel import Session, select

from src.db.billing import StripeSubscription
from src.db.membership_tiers import MembershipTier
from src.db.user_memberships import UserMembershipCreate
from src.services.membership_tiers.user_memberships import (
    assign_tier_to_user,
    auto_assign_free_tier,
)

logger = logging.getLogger(__name__)

# System user ID used when billing system modifies memberships
_SYSTEM_USER_ID = 0


# ---------------------------------------------------------------------------
# checkout.session.completed
# ---------------------------------------------------------------------------

def handle_checkout_completed(event: stripe.Event, db_session: Session) -> None:
    """
    On successful checkout:
    1. Create / update StripeSubscription record.
    2. Upgrade the user's membership tier.
    """
    session_obj = event.data.object  # type: ignore[attr-defined]

    stripe_subscription_id: str = session_obj.get("subscription")
    if not stripe_subscription_id:
        logger.warning("checkout.session.completed: no subscription id in event %s", event.id)
        return

    # Retrieve full subscription to get metadata and period dates
    stripe_sub = stripe.Subscription.retrieve(stripe_subscription_id)
    metadata: dict = stripe_sub.get("metadata", {})

    user_id_str = metadata.get("user_id")
    tier_id_str = metadata.get("tier_id")
    billing_period: str = metadata.get("billing_period", "monthly")

    if not user_id_str or not tier_id_str:
        logger.error(
            "checkout.session.completed: missing metadata in subscription %s",
            stripe_subscription_id,
        )
        return

    user_id = int(user_id_str)
    tier_id = int(tier_id_str)

    # Idempotency: skip if subscription record already exists
    existing = db_session.exec(
        select(StripeSubscription).where(
            StripeSubscription.stripe_subscription_id == stripe_subscription_id
        )
    ).first()

    if not existing:
        current_period = stripe_sub.get("current_period_start"), stripe_sub.get("current_period_end")
        record = StripeSubscription(
            user_id=user_id,
            tier_id=tier_id,
            stripe_subscription_id=stripe_subscription_id,
            status=stripe_sub.get("status", "active"),
            billing_period=billing_period,
            current_period_start=str(datetime.fromtimestamp(current_period[0])) if current_period[0] else None,
            current_period_end=str(datetime.fromtimestamp(current_period[1])) if current_period[1] else None,
            cancel_at_period_end=bool(stripe_sub.get("cancel_at_period_end", False)),
            creation_date=str(datetime.now()),
            update_date=str(datetime.now()),
        )
        db_session.add(record)
        db_session.commit()
        logger.info(
            "Created StripeSubscription for user_id=%s tier_id=%s sub=%s",
            user_id,
            tier_id,
            stripe_subscription_id,
        )
    else:
        logger.info(
            "StripeSubscription %s already exists — skipping create (idempotent)",
            stripe_subscription_id,
        )

    # Upgrade membership tier (assign_tier_to_user handles idempotency via status transitions)
    try:
        assign_tier_to_user(
            data=UserMembershipCreate(
                user_id=user_id,
                tier_id=tier_id,
                source="payment",
            ),
            assigned_by=_SYSTEM_USER_ID,
            db_session=db_session,
        )
        logger.info("Upgraded membership for user_id=%s to tier_id=%s", user_id, tier_id)
    except Exception as exc:
        logger.error("Failed to assign membership for user_id=%s: %s", user_id, exc)


# ---------------------------------------------------------------------------
# customer.subscription.deleted
# ---------------------------------------------------------------------------

def handle_subscription_deleted(event: stripe.Event, db_session: Session) -> None:
    """
    On subscription cancellation:
    1. Mark StripeSubscription as cancelled.
    2. Auto-assign the free tier.
    """
    stripe_sub = event.data.object  # type: ignore[attr-defined]
    stripe_subscription_id: str = stripe_sub.get("id", "")
    metadata: dict = stripe_sub.get("metadata", {})

    user_id_str = metadata.get("user_id")
    if not user_id_str:
        logger.error(
            "customer.subscription.deleted: no user_id in metadata for sub %s",
            stripe_subscription_id,
        )
        return

    user_id = int(user_id_str)

    record = db_session.exec(
        select(StripeSubscription).where(
            StripeSubscription.stripe_subscription_id == stripe_subscription_id
        )
    ).first()

    if record:
        record.status = "cancelled"
        record.update_date = str(datetime.now())
        db_session.add(record)
        db_session.commit()
        logger.info("Marked subscription %s as cancelled for user_id=%s", stripe_subscription_id, user_id)
    else:
        logger.warning(
            "customer.subscription.deleted: no local record found for sub %s",
            stripe_subscription_id,
        )

    # Downgrade to free tier
    try:
        auto_assign_free_tier(user_id, db_session)
        logger.info("Auto-assigned free tier for user_id=%s after subscription cancellation", user_id)
    except Exception as exc:
        logger.error("Failed to auto-assign free tier for user_id=%s: %s", user_id, exc)


# ---------------------------------------------------------------------------
# invoice.payment_failed
# ---------------------------------------------------------------------------

def handle_payment_failed(event: stripe.Event, db_session: Session) -> None:
    """
    On payment failure, mark the related subscription as past_due.
    """
    invoice = event.data.object  # type: ignore[attr-defined]
    stripe_subscription_id: str = invoice.get("subscription", "")

    if not stripe_subscription_id:
        logger.warning("invoice.payment_failed: no subscription id in event %s", event.id)
        return

    record = db_session.exec(
        select(StripeSubscription).where(
            StripeSubscription.stripe_subscription_id == stripe_subscription_id
        )
    ).first()

    if record:
        record.status = "past_due"
        record.update_date = str(datetime.now())
        db_session.add(record)
        db_session.commit()
        logger.info("Marked subscription %s as past_due", stripe_subscription_id)
    else:
        logger.warning(
            "invoice.payment_failed: no local record found for sub %s",
            stripe_subscription_id,
        )


# ---------------------------------------------------------------------------
# customer.subscription.updated
# ---------------------------------------------------------------------------

def handle_subscription_updated(event: stripe.Event, db_session: Session) -> None:
    """
    On subscription update, sync period dates and cancel_at_period_end flag.
    """
    stripe_sub = event.data.object  # type: ignore[attr-defined]
    stripe_subscription_id: str = stripe_sub.get("id", "")

    record = db_session.exec(
        select(StripeSubscription).where(
            StripeSubscription.stripe_subscription_id == stripe_subscription_id
        )
    ).first()

    if not record:
        logger.warning(
            "customer.subscription.updated: no local record found for sub %s",
            stripe_subscription_id,
        )
        return

    period_start = stripe_sub.get("current_period_start")
    period_end = stripe_sub.get("current_period_end")

    record.current_period_start = (
        str(datetime.fromtimestamp(period_start)) if period_start else record.current_period_start
    )
    record.current_period_end = (
        str(datetime.fromtimestamp(period_end)) if period_end else record.current_period_end
    )
    record.cancel_at_period_end = bool(stripe_sub.get("cancel_at_period_end", False))
    record.status = stripe_sub.get("status", record.status)
    record.update_date = str(datetime.now())

    db_session.add(record)
    db_session.commit()
    logger.info("Updated subscription record for %s", stripe_subscription_id)

File: src/endpoints/stripe/webhooks.ts
import type { Endpoint } from 'payload'
import { getStripe } from '../../stripe/client'
import {
  findTierByStripePrice,
  syncUserTier,
  downgradeToFree,
  matchPriceToInterval,
} from '../../hooks/syncUserTier'

export const stripeWebhookEndpoint: Endpoint = {
  path: '/stripe/webhooks',
  method: 'post',
  handler: async (req) => {
    const stripe = getStripe()

    // 1. Verify webhook signature
    const signature = req.headers.get('stripe-signature')
    const webhookSecret = process.env.STRIPE_WEBHOOK_SECRET

    if (!signature || !webhookSecret) {
      return Response.json({ error: 'Missing signature or webhook secret' }, { status: 400 })
    }

    let event
    try {
      const rawBody = await req.text?.()
      if (!rawBody) {
        return Response.json({ error: 'Empty request body' }, { status: 400 })
      }
      event = stripe.webhooks.constructEvent(rawBody, signature, webhookSecret)
    } catch (err) {
      console.error('[Stripe Webhook] Signature verification failed:', (err as Error).message)
      return Response.json({ error: 'Invalid signature' }, { status: 400 })
    }

    // 2. Idempotency check
    const existing = await req.payload.find({
      collection: 'stripe-events',
      where: { stripeEventId: { equals: event.id } },
      limit: 1,
      overrideAccess: true,
      req,
    })

    if (existing.totalDocs > 0) {
      return Response.json({ received: true, duplicate: true })
    }

    // 3. Process event
    let processed = true
    try {
      switch (event.type) {
        case 'checkout.session.completed':
          await handleCheckoutCompleted(event.data.object as any, req)
          break
        case 'customer.subscription.updated':
          await handleSubscriptionUpdated(event.data.object as any, req)
          break
        case 'customer.subscription.deleted':
          await handleSubscriptionDeleted(event.data.object as any, req)
          break
        case 'invoice.payment_failed':
          await handlePaymentFailed(event.data.object as any, req)
          break
        default:
          // Unhandled event type — ignore
          break
      }
    } catch (err) {
      console.error(`[Stripe Webhook] Error processing ${event.type}:`, (err as Error).message)
      processed = false
    }

    // 4. Log event for idempotency
    try {
      await req.payload.create({
        collection: 'stripe-events',
        data: {
          stripeEventId: event.id,
          eventType: event.type,
          processed,
        },
        overrideAccess: true,
        req,
      })
    } catch (err) {
      console.error('[Stripe Webhook] Failed to log event:', (err as Error).message)
    }

    // Always return 200 to Stripe
    return Response.json({ received: true })
  },
}

/**
 * Handle checkout.session.completed
 * Creates subscription record and upgrades user tier.
 */
async function handleCheckoutCompleted(session: any, req: any): Promise<void> {
  const userId = session.metadata?.userId
  const tierId = session.metadata?.tierId

  if (!userId || !tierId) {
    console.error('[Stripe Webhook] checkout.session.completed missing metadata')
    return
  }

  const stripe = getStripe()

  // Fetch the Stripe subscription for billing details
  const stripeSubscription = await stripe.subscriptions.retrieve(session.subscription as string)
  const priceId = stripeSubscription.items.data[0]?.price?.id
  const interval = stripeSubscription.items.data[0]?.plan?.interval

  // Find the tier to determine billing interval
  const tier = await req.payload.findByID({
    collection: 'membership-tiers',
    id: tierId,
    req,
    overrideAccess: true,
  })

  const billingInterval =
    matchPriceToInterval(tier, priceId ?? '') ?? (interval === 'year' ? 'yearly' : 'monthly')

  // Create subscription record
  await req.payload.create({
    collection: 'subscriptions',
    data: {
      user: userId,
      tier: tierId,
      stripeSubscriptionId: stripeSubscription.id,
      stripeCustomerId: session.customer as string,
      status: 'active',
      billingInterval,
      currentPeriodStart: new Date(
        stripeSubscription.current_period_start * 1000,
      ).toISOString(),
      currentPeriodEnd: new Date(stripeSubscription.current_period_end * 1000).toISOString(),
      cancelAtPeriodEnd: stripeSubscription.cancel_at_period_end,
    },
    overrideAccess: true,
    req,
  })

  // Upgrade user tier
  await syncUserTier(req.payload, req, userId, tierId)

  // Update any pending enrollments for this user's courses to 'paid'
  const pendingEnrollments = await req.payload.find({
    collection: 'enrollments',
    where: {
      and: [
        { user: { equals: userId } },
        { paymentStatus: { equals: 'pending' } },
      ],
    },
    overrideAccess: true,
    req,
  })

  for (const enrollment of pendingEnrollments.docs) {
    await req.payload.update({
      collection: 'enrollments',
      id: enrollment.id,
      data: { paymentStatus: 'paid' },
      overrideAccess: true,
      req,
    })
  }

  // Update stripeCustomerId on user if not set
  const user = await req.payload.findByID({
    collection: 'users',
    id: userId,
    req,
    overrideAccess: true,
  })

  if (!user.stripeCustomerId) {
    await req.payload.update({
      collection: 'users',
      id: userId,
      data: { stripeCustomerId: session.customer as string },
      req,
      overrideAccess: true,
    })
  }
}

/**
 * Handle customer.subscription.updated
 * Syncs subscription status and handles plan changes.
 */
async function handleSubscriptionUpdated(subscription: any, req: any): Promise<void> {
  // Find local subscription
  const localSubs = await req.payload.find({
    collection: 'subscriptions',
    where: { stripeSubscriptionId: { equals: subscription.id } },
    limit: 1,
    overrideAccess: true,
    req,
  })

  if (localSubs.totalDocs === 0) return

  const localSub = localSubs.docs[0]

  // Map Stripe status to our status
  let status = localSub.status
  if (subscription.status === 'active') status = 'active'
  else if (subscription.status === 'past_due') status = 'past_due'
  else if (subscription.status === 'canceled') status = 'cancelled'

  // Sync subscription data
  await req.payload.update({
    collection: 'subscriptions',
    id: localSub.id,
    data: {
      status,
      currentPeriodStart: new Date(subscription.current_period_start * 1000).toISOString(),
      currentPeriodEnd: new Date(subscription.current_period_end * 1000).toISOString(),
      cancelAtPeriodEnd: subscription.cancel_at_period_end,
    },
    overrideAccess: true,
    req,
  })

  // Check for plan change — if the price changed, update the user's tier
  const newPriceId = subscription.items?.data?.[0]?.price?.id
  if (newPriceId) {
    const newTier = await findTierByStripePrice(req.payload, req, newPriceId)
    if (newTier) {
      const currentTierId = typeof localSub.tier === 'string' ? localSub.tier : localSub.tier?.id
      if (newTier.id !== currentTierId) {
        // Plan changed — update subscription tier and user tier
        await req.payload.update({
          collection: 'subscriptions',
          id: localSub.id,
          data: { tier: newTier.id },
          overrideAccess: true,
          req,
        })
        const userId = typeof localSub.user === 'string' ? localSub.user : localSub.user?.id
        if (userId) {
          await syncUserTier(req.payload, req, userId, newTier.id)
        }
      }
    }
  }
}

/**
 * Handle customer.subscription.deleted
 * Downgrades user to free tier.
 */
async function handleSubscriptionDeleted(subscription: any, req: any): Promise<void> {
  const localSubs = await req.payload.find({
    collection: 'subscriptions',
    where: { stripeSubscriptionId: { equals: subscription.id } },
    limit: 1,
    overrideAccess: true,
    req,
  })

  if (localSubs.totalDocs === 0) return

  const localSub = localSubs.docs[0]

  // Mark subscription as cancelled
  await req.payload.update({
    collection: 'subscriptions',
    id: localSub.id,
    data: { status: 'cancelled' },
    overrideAccess: true,
    req,
  })

  // Downgrade user to free tier
  const userId = typeof localSub.user === 'string' ? localSub.user : localSub.user?.id
  if (userId) {
    await downgradeToFree(req.payload, req, userId)
  }
}

/**
 * Handle invoice.payment_failed
 * Marks subscription as past_due.
 */
async function handlePaymentFailed(invoice: any, req: any): Promise<void> {
  const subscriptionId = invoice.subscription as string
  if (!subscriptionId) return

  const localSubs = await req.payload.find({
    collection: 'subscriptions',
    where: { stripeSubscriptionId: { equals: subscriptionId } },
    limit: 1,
    overrideAccess: true,
    req,
  })

  if (localSubs.totalDocs === 0) return

  await req.payload.update({
    collection: 'subscriptions',
    id: localSubs.docs[0].id,
    data: { status: 'past_due' },
    overrideAccess: true,
    req,
  })
}
